meta_events.go

```diff
@@ -0,0 +1,15 @@
+package lightstep
+
+// keys for ls meta span events
+const LSMetaEvent_MetaEventKey = "lightstep.meta_event"
+const LSMetaEvent_PropagationFormatKey = "lightstep.propagation_format"
+const LSMetaEvent_TraceIdKey = "lightstep.trace_id"
+const LSMetaEvent_SpanIdKey = "lightstep.span_id"
+const LSMetaEvent_TracerGuidKey = "lightstep.tracer_guid"
+
+// operation names for ls meta span events
+const LSMetaEvent_ExtractOperation = "lightstep.extract_span"
+const LSMetaEvent_InjectOperation = "lightstep.inject_span"
+const LSMetaEvent_SpanStartOperation = "lightstep.span_start"
+const LSMetaEvent_SpanFinishOperation = "lightstep.span_finish"
```

Comment:
I don't see much value in a Finish meta event, since we just finished the actual span which includes all the same (and more) detail.

Reply:
My rationale with including start and finish is that you could easily determine how many unfinished spans exist via sets. It'd be a superset of spans, so you could determine "in-flight" spans more quickly?

Reply:
I can see some ways in which the analysis is more straightforward if we're simply subtracting finishes from starts. But, I feel that if we think about this in the form of relational queries, there's not much difference between selecting the finished span_ids from the stream of real spans vs selecting the finished span_ids from the stream of meta spans. https://lightstep.slack.com/archives/CFNR8UUNL/p1549575209243200

Reply:
I've reconsidered. Let's go ahead with a finish event.

Reply:
I was imagining a what-if scenario where we tried to forward these meta-spans to a separate lightstep instance, and to actually think of them as non-meta span data in this way. Then, I think, it would be nice if these meta spans formed meta-traces. This would require somehow coordinating so that all the meta-spans corresponding to a span have the same meta-trace_id. Can we do that?

Reply:
For a single process, it'd be possible (but probably kinda slow) - the tracer would need a map of meta-contexts to their associated live spans that would be referenced for every finish/inject.

The bigger problem is mostly, imo, RPCs. You could theoretically have inject reference that meta-context map then inject the meta-context as well as the live context (we'd need a different, new, header for this - something like `x-lightstep-meta-s/span/trace-id`) so that the associated meta-start event on the callee wouldn't break.

My thought would be that it'd be faster and easier to reassemble them on the remote side if we really wanted a 'trace' view of the meta events.

Reply:
I didn't mean for the meta "trace" to cross span boundaries, just one meta-trace per span. The meta trace-id could be derived by hashing the span_id of the span. Then if we directed this traffic to a lightstep instance of our own, we'd have a bunch of traces representing every span being sent.

Reply:
That'd be a pretty big change in scope since we'd need to have a separate code path to create the derivative trace id, then whatever hash function we're using would need to be ported across all the other tracer libraries.